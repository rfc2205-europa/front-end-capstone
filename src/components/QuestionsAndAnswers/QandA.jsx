import React from 'react';
import axios from 'axios';
import Search from './Search.jsx';
import QuestionsList from './QuestionsList/QuestionsList.jsx';
// import style from './style.css';


const service = axios.create({
  baseURL: 'http://localhost:3005',
  changeOrigin: true,
});

class QandA extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      questions: [],
      searchQ: [],
      smallView: true,
      searchView: false
    }
  }

  //fetches data from api on mounting of component
  componentDidMount = () => {
    this.fetch();
  }

  //fetches questions from api
  fetch = (product_id) => {
    var product_id = product_id || '66631';
    var body = { api: `https://app-hrsei-api.herokuapp.com/api/fec2/hr-rfc/qa/questions?product_id=${product_id}` }
    service.post('/retrieve', body)
      .then((res) => {
        console.log(res);
        this.setState({
          questions: res.data.results
        })
      })
      .catch((err) => {
        console.log('There is an error in fetching data: ', err)
      })
  }

  //searches state for matching questions
  search = (query, length) => {
    if (query.length > 2) {
      let searchResult = []
      this.state.questions.forEach((question) => {
        let body = question.question_body
        console.log('slice: ', body.slice(0, length), 'query: ', query);
        if (query === body.slice(0, length)) {
          console.log('match')
          searchResult.push(question)
        }
      })
      this.setState({
        searchQ: searchResult,
        searchView: true
      })
    } else {
      this.setState({
        searchView: false
      })
    }
  }

  addQs = () => {
    if(this.state.smallView){
      this.setState({
        smallView: false
      })
    }else{
      this.setState({
        smallView: true
      })
    }
  }

  render() {

    //conditional for rendering the search question list
    let listView;
    if (this.state.searchView) {
      listView = this.state.searchQ
    } else {
      listView = this.state.questions
    }

    let button;
    if (this.state.smallView) {
      if (this.state.questions.length > 2) {
        button = <button onClick={this.addQs}>Add More Answered Questions</button>
      } else {
        button = <div></div>
      }
    } else {
      button = <button onClick={this.addQs}>Remove Added Questions and Answers</button>
    }


    return (
      <div className="q_and_a">
        <h3>Questions and Answers</h3>
        <Search search={this.search} questions={this.state.questions} />
        <div className="qList">
          <QuestionsList questions={listView} smallV={this.state.smallView}fetch={this.fetch} />
        </div>
        {button}<button>Add a Question</button>
      </div>
    )
  }
}

export default QandA;
