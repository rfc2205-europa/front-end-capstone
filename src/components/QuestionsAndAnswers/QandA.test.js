import renderer from 'react-test-renderer';
import QandA from './QandA.jsx';
import React from 'react';


describe("QandA Component Testing", function () {
  it('renders the QandA component correctly', () => {
    const component = renderer.create(<QandA />);
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot();
  });

});
