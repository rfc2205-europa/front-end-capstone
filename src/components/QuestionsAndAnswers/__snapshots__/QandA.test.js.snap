// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QandA Component Testing renders the QandA component correctly 1`] = `
<div>
  <h3>
    Questions and Answers
  </h3>
  <div
    className="search"
  >
    <form>
      <input
        onChange={[Function]}
        placeholder="Have a Question Search For Answers"
        type="text"
      />
      <input
        type="submit"
      />
    </form>
  </div>
  <div>
    <div>
      <div
        className="tile"
      >
        <div
          className="first_line"
        >
          <h3>
            Q:
          </h3>
          <h5
            onClick={[Function]}
          >
            Helpful? Yes 
          </h5>
          <button>
            add answer
          </button>
        </div>
        <div>
          A:
        </div>
      </div>
    </div>
    <button
      onClick={[Function]}
    />
    <button>
      Add A Question
    </button>
  </div>
</div>
`;
